### PR TITLE
tools: periodically clean up bare-metal servers

### PR DESCRIPTION
**Pull request by davidweisse**
This adds a cronjob to clean up the runtime classes on the bare-metal servers. All unused runtime classes get deleted and the containerd config is updated, and the corresponding folder in `/opt/edgeless` gets deleted as well.

### CODE REVIEW

File: packages/cleanup-bm.sh [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+# SPDX-License-Identifier: AGPL-3.0-only
+
+set -euo pipefail
+
+echo "Starting cleanup"
+
+mapfile -t runtimeClasses < <(kubectl get runtimeclass -o jsonpath='{.items[*].metadata.name}' | tr ' ' '\n' | grep '^contrast-cc')
+
+declare -a usedRuntimeClasses=()
+declare -a unusedRuntimeClasses=()
```
Comment (burgerdev) [inline packages/cleanup-bm.sh:12]: I find the logic in this script hard to follow. Have you considered just filling a temp file with runtimeclasses in use, and using `comm(1)` to do the set difference?

File: packages/cleanup-bm.sh [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+resourcesToCheck=(
```
Comment (burgerdev) [inline packages/cleanup-bm.sh:14]: If we checked for an existing node installer, we could tell whether a runtime is just about to be installed.

File: packages/cleanup-bm.sh [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  # Delete unused runtime classes
+  echo "Deleting runtimeclass ${runtimeClass} ..."
+  kubectl delete runtimeclass "${runtimeClass}"
+
+  # Delete unused binaries
+  path="${OPTEDGELESS}/${runtimeClass}"
+  if [ -d "${path}" ]; then
+    echo "Deleting binaries from ${OPTEDGELESS}/${runtimeClass} ..."
+    rm -rf "${path}"
+  fi
```
Comment (Freax13) [inline packages/cleanup-bm.sh:53]: We also store data in `/var/lib/nydus-snapshotter/`.

File: packages/containers.nix [hunk 4]
```diff
@@ -155,6 +155,19 @@ let
...
       };
     };
+
+    cleanup-bm = dockerTools.buildImage {
+      name = "cleanup-bm";
+      tag = "v0.0.1";
+      copyToRoot = with pkgs; [
+        busybox
+        kubectl
+        dasel
```
Comment (Freax13) [inline packages/containers.nix:165]: Do we need this here or is it enough to specify this in `writeShellApplication`?

File: packages/cleanup-bm.sh (hunk 2 above)
Comment (davidweisse) [reply packages/cleanup-bm.sh:14]: Is there any way of associating a given node installer with a runtime handler, so we can mark this handler as used?

Comment (burgerdev) [reply packages/cleanup-bm.sh:14]: Given a `RuntimeClass`, you'd need to check the args of the node-installer containers. We could consider making this relationship easier to discover, e.g. by adding a label.

File: packages/cleanup-bm.sh [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+for runtimeClass in "${unusedRuntimeClasses[@]}"; do
+  # Delete unused runtime classes
+  echo "Deleting runtimeclass ${runtimeClass} ..."
+  kubectl delete runtimeclass "${runtimeClass}"
+
+  # Delete unused files
+  if [ -d "${OPTEDGELESS}/${runtimeClass}" ]; then
+    echo "Deleting files from ${OPTEDGELESS}/${runtimeClass} ..."
+    rm -rf "${OPTEDGELESS:?}/${runtimeClass}"
```
[the remark this answers is not in the document]
Comment (davidweisse) [reply packages/cleanup-bm.sh:47]: `nix fmt` seemed to complain about `OPTEDGELESS` possibly expanding to `/`, but in the other cases that would not be a problem.

**Review by Freax13 — APPROVED** [review]

**Review by burgerdev — APPROVED** [review]
